### PR TITLE
Make `_embedding_bag_backward` explicitly dispatch to CPU and CUDA.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129691

This PR modifies `_embedding_bag_backward` item inside _native_functions.yaml_, so that it
dispatches to CPU and CUDA directly, instead of `CompositeImplicitAutograd`.

*Context:* PyTorch operations that have the `CompositeImplicitAutograd` dispatch do not
allow third party backends (e.g. XLA) to modify its implementation, since this dispatch
key has higher priority. When calling `_embedding_bag_backward` operation using XLA, a
dispatch error will be thrown, since PyTorch/XLA doesn't support sparse tensors.

*Problem:* `_embedding_bag_backward` has a `sparse` parameter that controls whether the
operation should return a sparse or dense tensor. However, at the moment, PyTorch/XLA does
not support sparse tensors. In order to fallback that execution to dense, i.e. change the
flag at runtime, we need to be able to modify its implementation.

*Solution:* we have changed the dispatch of `_embedding_bag_backward` to CPU and CUDA,
which allowed us to introduce our own kernel for it.

Additionally, this PR refactored the representation of its mode from constant integers
into an enum class. It also introduces two additional operators: `int == EmbeddingBagMode`
and `int != EmbeddingBagMode`.

cc @bdhirsh @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @miladm @lezcano

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1998,6 +1998,13 @@
   weight: _embedding_bag_backward_symint(grad, indices, offsets, result1, result2, result3, weight.sym_size(0), scale_grad_by_freq, mode, sparse, per_sample_weights, padding_idx)
   per_sample_weights: _embedding_bag_per_sample_weights_backward(grad, weight, indices, offsets, result1, mode, padding_idx)
 
+- name: _embedding_bag_backward(Tensor grad, Tensor indices, Tensor offsets, Tensor offset2bag, Tensor bag_size, Tensor maximum_indices, SymInt num_weights, bool scale_grad_by_freq, int mode, bool sparse, Tensor? per_sample_weights, int padding_idx=-1) -> Tensor
+  indices: non_differentiable
```
Comment (lezcano) [inline tools/autograd/derivatives.yaml:2002]: add that grad is `not_implemented`

File: aten/src/ATen/native/cuda/EmbeddingBag.cu [hunk 2]
```diff
@@ -450,16 +447,16 @@ Tensor _embedding_bag_dense_backward_cuda(const Tensor &grad_, const Tensor &ind
   checkSameGPU("embedding_bag_cuda", grad_arg, indices_arg);
 
 
-  switch (mode) {
-    case MODE_SUM:
-    case MODE_MEAN:
-      if (mode == MODE_MEAN)
+  switch (static_cast<EmbeddingBagMode>(mode)) {
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/EmbeddingBag.cu:450]: if you want to go for the full refactor, you should probably never have static_casts, and always pass the right type around.

**Review by lezcano — APPROVED** [review]

**Comment by lezcano** [comment]
it seems you'll also need to change a few other things here and there.

File: aten/src/ATen/native/cuda/EmbeddingBag.cu (hunk 2 above)
Comment (ysiraichi) [reply aten/src/ATen/native/cuda/EmbeddingBag.cu:450]: I wasn't really planning on going all the way, at least in this PR. This refactor was just to centralize the mode enum (we were declaring constant ints in 2 places already).

**Comment by ysiraichi** [comment]
@lezcano Incorporated your suggestions + a few other changes to fix CI. Let me know if this is good to go!

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -2021,6 +2021,14 @@
   weight: _embedding_bag_backward_symint(grad, indices, offsets, result1, result2, result3, weight.sym_size(0), scale_grad_by_freq, mode, sparse, per_sample_weights, padding_idx)
   per_sample_weights: _embedding_bag_per_sample_weights_backward(grad, weight, indices, offsets, result1, mode, padding_idx)
 
+- name: _embedding_bag_backward(Tensor grad, Tensor indices, Tensor offsets, Tensor offset2bag, Tensor bag_size, Tensor maximum_indices, SymInt num_weights, bool scale_grad_by_freq, int mode, bool sparse, Tensor? per_sample_weights, int padding_idx=-1) -> Tensor
+  grad: non_differentiable
```
Comment (lezcano) [inline tools/autograd/derivatives.yaml:2025]: it is differentiable, it's `not_ipmlemented` though.

**Review by lezcano — APPROVED** [review]
Just a minor point, otherwise LGTM

**Comment by ysiraichi** [comment]
These CI failures don't seem related to this PR, since they are failing on other commits, too. I will merge this PR with `-f`.

**Pull request closed without merging**